File: clients/java/src/main/java/io/weedforge/FileId.java
package io.weedforge;

/**
 * A SeaweedFS file identifier: a first-class value, not an opaque string.
 *
 * <p>The wire form is {@code {volumeId},{hex}} where {@code hex} encodes the 12-byte
 * big-endian buffer {@code [needleKey(8) | cookie(4)]} with leading zero BYTES stripped.
 * The needle key is a full 64-bit value (NOT 32-bit), so this type represents the entire
 * SeaweedFS fid space. All three components are stored as {@code long} and treated as
 * unsigned bit patterns (Java has no unsigned primitives).
 */
public final class FileId {

    private static final int COOKIE_HEX_LEN = 8;   // cookie = 4 bytes -> 8 hex chars
    private static final int MAX_KC_HEX_LEN = 24;  // 8-byte key + 4-byte cookie

    private final long volumeId; // unsigned u32
    private final long fileKey;  // unsigned u64 bit pattern
    private final long cookie;   // unsigned u32

    public FileId(long volumeId, long fileKey, long cookie) {
        this.volumeId = volumeId;
        this.fileKey = fileKey;
        this.cookie = cookie;
    }

    /** @return the volume id (unsigned 32-bit, held in a long). */
    public long volumeId() { return volumeId; }

    /** @return the needle/file key (unsigned 64-bit bit pattern). */
    public long fileKey() { return fileKey; }

    /** @return the cookie (unsigned 32-bit, held in a long). */
    public long cookie() { return cookie; }

    /**
     * Parses a SeaweedFS fid string. Mirrors SeaweedFS's own parser: the last 8 hex
     * characters are the cookie, and the remaining prefix is the (up to 64-bit) needle key.
     *
     * @throws WeedException.InvalidFileId if the string is malformed.
     */
    public static FileId parse(String s) {
        int comma = s.indexOf(',');
        if (comma < 0) {
            throw new WeedException.InvalidFileId(s, "missing comma separator");
        }
        String volStr = s.substring(0, comma);
        String kc = s.substring(comma + 1);

        if (!isAllDigits(volStr)) {
            throw new WeedException.InvalidFileId(s, "invalid volume id");
        }
        long vol;
        try {
            vol = Long.parseLong(volStr);
        } catch (NumberFormatException e) {
            throw new WeedException.InvalidFileId(s, "invalid volume id: " + e.getMessage());
        }
        if (vol < 0 || vol > 0xFFFFFFFFL) {
            throw new WeedException.InvalidFileId(s, "volume id out of range");
        }

        // SeaweedFS requires len(kc) > 8 and <= 24. Rejecting all non-hex bytes here also
        // rejects a leading '+'/'-' that Long.parseUnsignedLong would otherwise accept.
        if (kc.length() <= COOKIE_HEX_LEN || kc.length() > MAX_KC_HEX_LEN) {
            throw new WeedException.InvalidFileId(s, "key/cookie hex length out of range");
        }
        if (!isAllHex(kc)) {
            throw new WeedException.InvalidFileId(s, "non-hex character in key/cookie");
        }

        int split = kc.length() - COOKIE_HEX_LEN;
        String keyHex = kc.substring(0, split);
        String cookieHex = kc.substring(split);

        long key;
        long ck;
        try {
            key = Long.parseUnsignedLong(keyHex, 16); // up to 16 hex -> full 64 bits
            ck = Long.parseLong(cookieHex, 16);       // exactly 8 hex -> fits a signed long
        } catch (NumberFormatException e) {
            throw new WeedException.InvalidFileId(s, "invalid key/cookie hex: " + e.getMessage());
        }
        return new FileId(vol, key, ck);
    }

    /**
     * Returns the canonical SeaweedFS string form (leading zero bytes stripped).
     * {@code render(parse(s))} reproduces the server's representation; {@code parse(render(f)).equals(f)}.
     */
    public String render() {
        byte[] buf = new byte[12];
        for (int i = 0; i < 8; i++) {
            buf[i] = (byte) (fileKey >>> (56 - 8 * i));
        }
        for (int i = 0; i < 4; i++) {
            buf[8 + i] = (byte) (cookie >>> (24 - 8 * i));
        }
        int start = 0;
        while (start < buf.length - 1 && buf[start] == 0) {
            start++;
        }
        StringBuilder sb = new StringBuilder(32);
        sb.append(Long.toUnsignedString(volumeId)).append(',');
        for (int i = start; i < buf.length; i++) {
            int b = buf[i] & 0xFF;
            sb.append(Character.forDigit(b >>> 4, 16));
            sb.append(Character.forDigit(b & 0xF, 16));
        }
        return sb.toString();
    }

    @Override
    public String toString() { return render(); }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof FileId)) return false;
        FileId f = (FileId) o;
        return volumeId == f.volumeId && fileKey == f.fileKey && cookie == f.cookie;
    }

    @Override
    public int hashCode() {
        int result = Long.hashCode(volumeId);
        result = 31 * result + Long.hashCode(fileKey);
        result = 31 * result + Long.hashCode(cookie);
        return result;
    }

    private static boolean isAllDigits(String s) {
        if (s.isEmpty()) return false;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c < '0' || c > '9') return false;
        }
        return true;
    }

    private static boolean isAllHex(String s) {
        if (s.isEmpty()) return false;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            boolean hex = (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
            if (!hex) return false;
        }
        return true;
    }
}

File: clients/go/fileid.go
package weedforge

import (
	"encoding/binary"
	"encoding/hex"
	"strconv"
	"strings"
)

// FileId is a SeaweedFS file identifier: a first-class value, not an opaque string.
//
// The wire form is "{volumeId},{hex}" where hex encodes the 12-byte big-endian
// buffer [needleKey(8) | cookie(4)] with leading zero BYTES stripped. The needle
// key is a full 64-bit value (NOT 32-bit), so this type represents the entire
// SeaweedFS fid space.
type FileId struct {
	VolumeID uint32
	FileKey  uint64
	Cookie   uint32
}

// NewFileId builds a FileId from its components.
func NewFileId(volumeID uint32, fileKey uint64, cookie uint32) FileId {
	return FileId{VolumeID: volumeID, FileKey: fileKey, Cookie: cookie}
}

const (
	cookieHexLen = 8  // cookie is 4 bytes -> 8 hex chars
	maxKCHexLen  = 24 // 8-byte key + 4-byte cookie -> 24 hex chars
)

// ParseFileID parses a SeaweedFS fid string. It mirrors SeaweedFS's own parser:
// the last 8 hex characters are the cookie and the remaining prefix is the
// (up to 64-bit) needle key.
func ParseFileID(s string) (FileId, error) {
	comma := strings.IndexByte(s, ',')
	if comma < 0 {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "missing comma separator"}
	}
	volStr, kc := s[:comma], s[comma+1:]

	if !isAllDigits(volStr) {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "invalid volume id"}
	}
	vol, err := strconv.ParseUint(volStr, 10, 32)
	if err != nil {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "invalid volume id: " + err.Error()}
	}

	// SeaweedFS requires len(kc) > 8 (must carry at least the cookie plus one key
	// nibble) and <= 24. Rejecting all non-hex bytes here also rejects '+'/'-' signs
	// that strconv.ParseUint would otherwise silently accept.
	if len(kc) <= cookieHexLen || len(kc) > maxKCHexLen {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "key/cookie hex length out of range"}
	}
	if !isAllHex(kc) {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "non-hex character in key/cookie"}
	}

	split := len(kc) - cookieHexLen
	keyHex, cookieHex := kc[:split], kc[split:]

	key, err := strconv.ParseUint(keyHex, 16, 64)
	if err != nil {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "invalid needle key: " + err.Error()}
	}
	cookie, err := strconv.ParseUint(cookieHex, 16, 32)
	if err != nil {
		return FileId{}, &InvalidFileIDError{Value: s, Reason: "invalid cookie: " + err.Error()}
	}

	return FileId{VolumeID: uint32(vol), FileKey: key, Cookie: uint32(cookie)}, nil
}

// Render returns the canonical SeaweedFS string form (leading zero bytes stripped).
// Render(Parse(s)) reproduces the server's representation; Parse(Render(f)) == f.
func (f FileId) Render() string {
	var buf [12]byte
	binary.BigEndian.PutUint64(buf[0:8], f.FileKey)
	binary.BigEndian.PutUint32(buf[8:12], f.Cookie)

	// Strip leading zero bytes, keeping at least the final byte (so the all-zero
	// null fid renders as "00" rather than empty).
	i := 0
	for i < len(buf)-1 && buf[i] == 0 {
		i++
	}
	return strconv.FormatUint(uint64(f.VolumeID), 10) + "," + hex.EncodeToString(buf[i:])
}

// String implements fmt.Stringer.
func (f FileId) String() string { return f.Render() }

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

func isAllHex(s string) bool {
	if s == "" {
		return false
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		isHex := (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
		if !isHex {
			return false
		}
	}
	return true
}
